[options] adding exceptions to max-line-length rule (closes #1495)

jmlopez-rod opened the pull request:
Adding `ignoreUrls`, `ignoreImports`, and `ignoreComments` options to
the rule.

JKillian commented on test/rules/max-line-length/ignore-imports/test.ts.lint:
```diff
@@ -0,0 +1,29 @@
+
+import {
+    obj1, obj2, obj3, obj4, just, trying, to, be, a, rebel, here
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [Exceeds maximum line length of 50]
```
shouldn't this be okay?

jmlopez-rod replied:
This shouldn't be ignored. If they choose to do

```
import { obj1, obj2, obj3, obj4, just, trying, to, be, a, rebel, here } from '...';
```

then that is ok, but the moment they decide to break the line then we have options to prevent exceeding the line limit. I usually prefer to place each import on its own line when using this form:


```
import {
  obj1,
  obj2,
  obj3,
  obj4,
  just,
  trying,
  to,
  be,
  a,
  rebel,
  here,
} from '...';
```

I decided to go with this approach because in the past I have seen a really long list increasing on the same line and it gets hard to see the changes. So really the rule should just apply to the string part of the import which we don't have much control over.

JKillian commented on src/rules/maxLineLengthRule.ts:
```diff
@@ -19,6 +19,8 @@ import * as ts from "typescript";
 
+import { IDisabledInterval } from "../language/rule/rule";
+
 export class Rule extends Lint.Rules.AbstractRule {
```
Can we declare the rule options as constants, and then use those constants in the rule metadata and in the code? 

(For example, [similar to this](https://github.com/palantir/tslint/blob/master/src/rules/tripleEqualsRule.ts#L22L51))

jmlopez-rod replied:
Ah, that can happen. Got it, will apply the change.

JKillian commented on src/rules/maxLineLengthRule.ts:
```diff
@@ -19,6 +19,8 @@ import * as ts from "typescript";
+
 export class Rule extends Lint.Rules.AbstractRule {
     /* tslint:disable:object-literal-sort-keys */
     public static metadata: Lint.IRuleMetadata = {
```
Let's update this as well - this info is used to generate the docs website when a release is published

JKillian commented on src/rules/maxLineLengthRule.ts:
```diff
@@ -71,11 +76,73 @@ export class Rule extends Lint.Rules.AbstractRule {
                 if (!Lint.doesIntersect(ruleFailure, disabledIntervals)) {
-                    ruleFailures.push(ruleFailure);
+                    let report = true;
+                    if (ruleArgs.indexOf("ignoreUrls") > -1) {
```
You can just use `this.hasOption(OPTION_IGNORE_URLS)` here

jmlopez-rod replied:
I think this can only be used on a walker. I was going to move the logic in the `apply` function to the walker so I'll be able to do this there.

jmlopez-rod commented:
@JKillian , I have refactored the rule so that it may be easier to add other rules in the future. That is, in case we want to add other eslint options that I'm not adding such as disabling the rule in strings and templates. The only other thing I added was support to ignore a user defined pattern. This is particularly useful when using `require` (another form of imports). I have added tests for this as well.

I hope the update to the documentation was done properly.

jmlopez-rod commented:
@JKillian I just noticed that the current code as it stands including my PR does not cover the case when users use tabs instead of spaces. That is we are assuming that `\t` counts for 1 character. In eslint they have a `tabWidth` option which defaults to 4 to properly account for tab use. I bring this up mainly because I would like to push for getting rid of the arguments and instead just use an object to declare the exceptions. Obviously since the original rule only has one argument we must leave it as it. But once we add `ignore-urls`, `ignore-imports` and `ignore-comments` as arguments we will have to keep them and the rule schema will just get larger. Thoughts?

adidahiya commented:
@JKillian can you take another look this week?

jmlopez-rod commented:
Hi guys, this PR has diverted a bit. I have released this rule in the tslint-eslint-rules repo (https://github.com/buzinas/tslint-eslint-rules/blob/master/src/docs/rules/terMaxLenRule.md). The rule does the same as the one in eslint. I'm closing this PR to help you clean up. If you guys want  to make improvements to the rule you peek into the source file. Thank you for your comments @JKillian , I learned a few things things from them.

JKillian commented:
Apologies @jmlopez-rod for not not fully seeing this through the whole way, thanks for the understanding! I haven't had much extra time to work on TSLint recently (although @nchen63 has been doing a great job!). Glad you got things merged over in `tslint-eslint-rules`!
